pkg/resources/helpers.go

```diff
@@ -142,3 +146,77 @@ func IsDataType() schema.SchemaValidateFunc { //nolint:staticcheck
 		return warnings, errors
 	}
 }
+
+// IsValidIdentifier is a validator that can be used for validating identifiers passed in resources and data sources.
+// Typically, we expect passed identifiers to be a variation of sdk.ObjectIdentifier. To use this function, pass it as
+// a validation function on identifier field with generic parameter set to the desired sdk.ObjectIdentifier implementation.
+func IsValidIdentifier[T sdk.ObjectIdentifier]() schema.SchemaValidateDiagFunc {
+	return func(value any, path cty.Path) diag.Diagnostics {
+		var diags diag.Diagnostics
+
+		// For now, we won't support sdk.ExternalObjectIdentifiers. The reason behind it is that the functions that parse identifiers are not
+		// able to differentiate between sdk.ExternalObjectIdentifiers and sdk.DatabaseObjectIdentifier or sdk.SchemaObjectIdentifier,
+		// because sdk.ExternalObjectIdentifiers has varying parts count (2 or 3).
+		if _, ok := any(sdk.ExternalObjectIdentifier{}).(T); ok {
+			diags = append(diags, diag.Diagnostic{
+				Severity:      diag.Error,
+				Summary:       "Invalid schema identifier type",
+				Detail:        "Identifier validation is not available for sdk.ExternalObjectIdentifier type. This is a provider error please file a report: https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/new/choose",
+				AttributePath: path,
+			})
+			return diags
+		}
+
+		if stringValue, ok := value.(string); ok {
+			id, err := helpers.DecodeSnowflakeParameterID(stringValue)
+			if err != nil {
+				diags = append(diags, diag.Diagnostic{
+					Severity: diag.Error,
+					Summary:  "Unable to parse the identifier",
+					Detail: fmt.Sprintf(
+						"Unable to parse the identifier: %s. Make sure you are using the correct form of the fully qualified name for this field: %s",
+						stringValue,
+						getExpectedIdentifierForm[T](nil),
+					),
+					AttributePath: path,
+				})
+				return diags
+			}
+
+			if _, ok := id.(T); !ok {
+				diags = append(diags, diag.Diagnostic{
+					Severity: diag.Error,
+					Summary:  "Invalid identifier type",
+					Detail: fmt.Sprintf(
+						"Expected %s identifier type, but got: %T. The correct form of the fully qualified name for this field is: %s, but was %s",
+						reflect.TypeOf(new(T)).Elem().Name(),
+						id,
+						getExpectedIdentifierForm[T](nil),
+						getExpectedIdentifierForm(&id),
+					),
+					AttributePath: path,
+				})
+			}
+		} else {
+			diags = append(diags, diag.Diagnostic{
+				Severity:      diag.Error,
+				Summary:       "Invalid schema identifier type",
+				Detail:        fmt.Sprintf("Expected schema string type, but got: %T. This is a provider error please file a report: https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/new/choose", value),
```

Comment:
why do we want to consider this a provider error? This just means that it is not used on string

Reply:
It's a provider error because this 'if' branch means that the identifier from the schema is not a string (e.g. could be int as Terraform type). Anyway, it would later fail on helpers.DecodeSnowflakeParameterID.